File: framework-learning/reactWorkspace/react-course-v16/src/examples/This.js
import React from 'react';

class App extends React.Component {
    constructor(props) {
        super(props);
        // 像这种如果多个状态都依赖于该方法，最好像下面一样把他提取出来
        // this.state = {
        //     person: this.caculatePerson(),
        //     age: this.caculatePerson().age,
        // }
        const person = this.caculatePerson();
        this.state = {
            person,
            age: person.age,
        }
    }
    caculatePerson() {
        return {
            name: 'gao',
            age: 23,
        };
    }
    handleClick(e) {
        console.log(this)
        console.log(App.prototype)
        e.preventDefault();
        console.log('The link was clicked.');
    }
    
    render() {
        return (
            <a href="" onClick={this.handleClick.bind(this)}>
                Click me
            </a>
        );
    }
}

// functional component
// function App() {
    // function handleClick(e) {
    //     e.preventDefault();
    //     console.log('The link was clicked.');
    // }
  
    // return (
    //     <a href="#" onClick={handleClick}>
    //         Click me
    //     </a>
    // );
// }

console.log(App.prototype)
export default App;

File: framework-learning/c++Workspace/core/class/this.cpp
#include <iostream>
using namespace std;
#include <string>

class Person {
    public:
        Person(int age) {
            // this 指针指向被调用的成员函数所属的对象
            this->age = age;
        }
        int getAge() {
            return this->age;
        }
         // 以引用的方式返回，每次不会生成一个新的对象，因此结果是 30
        Person& personAddAge(Person &p) {
            this->age += p.age;
            return *this;
        }
        // 以值的方式返回，每次会生成一个新的对象，因此结果是 20
        // Person personAddAge(Person &p) {
        //     this->age += p.age;
        //     return *this;
        // }
    private:
        int age;
};

int main() {
    Person p1 = Person(10);
    Person p2 = Person(10);
    p1.personAddAge(p2).personAddAge(p2);

    cout << p1.getAge() << endl;

    return 0;
}